pkg/controller/configmap/configmap.go

```diff
@@ -0,0 +1,211 @@
+	reqLogger logr.Logger,
+	cr *corev1.ConfigMap,
+) error {
+	// Memoize state of the incoming object.
+	oldMap, _ := observer.GetConfigMap(cr.Namespace, cr.Name)
+
+	// Set a defer func to write new status and/or finalizers if they change.
+	defer func() {
+
+		if _, ok := oldMap.Labels["kubedirectorcmtype"]; !ok {
+			return
+		}
+		/* anonymous fun to check if some cluster
+		   is using this config map as a connection */
+		isClusterUsingConfigMap := func(cmName string, cluster kdv1.KubeDirectorCluster) bool {
+			clusterModels := cluster.Spec.Connections.ConfigMaps
+			for _, modelMapName := range clusterModels {
+				if modelMapName == cmName {
+					return true
+				}
+			}
+			return false
+		}
+		allClusters := &kdv1.KubeDirectorClusterList{}
+		shared.List(context.TODO(), allClusters)
+		for _, kubecluster := range allClusters.Items {
+			if isClusterUsingConfigMap(cr.Name, kubecluster) {
+				shared.LogInfof(
+					reqLogger,
+					cr,
+					shared.EventReasonConfigMap,
+					"configmap {%s} is connected to cluster {%s} updating its configmeta",
+					cr.Name,
+					kubecluster.Name,
+				)
+				updateMetaGenerator := &kubecluster
+				updateMetaGenerator.Spec.ConfigMetaGenerator = kubecluster.Spec.ConfigMetaGenerator + 1
+				//Notify cluster by incrementing configmetaGenerator
+				wait := time.Second
+				maxWait := 4096 * time.Second
+				for {
+					if shared.Update(context.TODO(), updateMetaGenerator) == nil {
+						break
+					}
+					if wait > maxWait {
+						shared.LogErrorf(
+							reqLogger,
+							fmt.Errorf("failed to update cluster"),
+							cr,
+							shared.EventReasonConfigMap,
+							"Unable to notify cluster {%s} of configmeta change",
+							updateMetaGenerator.Name)
+						break
+					}
+					time.Sleep(wait)
+					wait = wait * 2
+				}
+			}
+		}
+	}()
+
+	// We use a finalizer to maintain config state consistency.
+	doExit, finalizerErr := r.handleFinalizers(reqLogger, cr)
+	if finalizerErr != nil {
+		return finalizerErr
+	}
+	if doExit {
+		return nil
+	}
+
+	// For a new KD config just update the status state/gen.
+	shouldProcessCR, processErr := r.handleNewConfigMap(reqLogger, cr)
+	if processErr != nil {
+		return processErr
+	}
+	if !shouldProcessCR {
+		return nil
+	}
+	return nil
+}
+
+// handleNewConfig looks in the cache for the last-known status generation
+// UID for this CR. If there is one, make sure the UID is what we expect, and
+// if so return true to keep processing the CR. If there is not any last-known
+// UID, this is either a new CR or one that was created before this KD came up.
+// In the former case, where the CR status itself has no generation UID: set
+// the config state to creating (this will also trigger population of the
+// generation UID) and return false to cause this handler to exit; we'll pick
+// up further processing in the next handler. In the latter case, sync up our
+// internal state with the visible state of the CR and return true to continue
+// processing.
```

Comment:
These comments about status generation don't apply here. Also in general I don't know that a function like this serves any purpose for what we are doing with configmaps. There's a bunch of things in the function that puzzle me but my top-level reaction would be that we don't need it.